### DISCUSSION TITLE
Proposal: have an option to declare frames using coordinates (x, y) at frame center instead of top-left corner.

### DISCUSSION THREAD

**Discussion by jodros** [sile-typesetter/sile#1944] [Ideas]
I know we can just do

```lua
frame = {
   top = y,
   left = x,
   width = w,
   height = h
}
```

But in this case (x, y) will always point to the top-left corner. Personally, I prefer that coordinates could point to the frame center as well.

I'm afraid about how naive my approach is, but I couldn't think how to handle `width` and `height` options and the centered coordinates at the same time: 

```lua
return function()
    local plain = require "classes.plain"
    local x, y = "8cm", "9cm"
    local w, h = "2cm", "4cm"

    plain.defaultFrameset = {
        content = {
            left = x .. '-(' .. w .. "/2)",
            right = x .. "+(" .. w .. "/2)",
            top = y .. '-(' .. h .. "/2)",
            bottom = y .. "+(" .. h .. "/2)"
        }
    }
    local class = plain()
    SILE.documentState.documentClass = class
end
```

**Comment by Omikhleia** [comment]
I am afraid that I don't understand the idea here (nor the hard-coded values in the example, regardless of the paper size). Is there something you can't achieve with the existing frame definition logic, or that can't be made with a (thin) layer on top of it?

**Reply by jodros** [reply]
Actually, there is nothing I can't achieve with the existing frame definition.

The thing is that I find easier to visualize mentally how the frames will look when I declare them this way, and suppose I'm not the only one who finds so. My idea is just to add the option of declaring this way too.

> nor the hard-coded values in the example

Well, here is an explanation:

The first example, both $width$ and $height$ have their start from $left$ and $top$.
![изображение](https://github.com/sile-typesetter/sile/assets/15092760/0db8e183-fe90-4910-8d90-a2b592285e00)

But what if we just want $(x, y)$ to be at the same time:
- Like the Cartesian origin relatively to the frame itself 
- Like any Cartesian point relatively to the page

In order to do so, I could think only in dividing $width$ and $height$ by half, so each half of $height$ would correspond either to  $top$ or $bottom$ and so for the $width$, then the constraints may be calculated as:

- The $left$ is the point of origin of the frame less half of its width: $left = x - \left(\frac w 2\right)$
- The $right$ is the point of origin of the frame plus half of its width: $right = x + \left(\frac w 2\right)$

![изображение](https://github.com/sile-typesetter/sile/assets/15092760/770dd875-95a7-455e-9f20-ab500e200f48)

I hope this explanation is enough.

> or that can't be made with a (thin) layer on top of it?

That's exactly what I'm doing here, I just want an opinion whether there is a smarter way of doing so, and eventually adding it to SILE. I believe some people out there would also like to declare frames this way:

```lua
frame = {
   y = "",
   x = "",
   width = "",
   height = ""
}
```

Maybe not too many people, but why not to just have the option?

**Comment by Omikhleia** [comment]
> I hope this explanation is enough.

Thanks for the explanation, I think I got it!

_As far as I know,_ it's quite unusual in page layout design to refer to the center of the page -- most page layouts I've heard of are usually defined by the text block's dimensions and the margins, often in some relative arithmetic proportions (for instance, most page layouts mentioned by Bringhurst in his _Elements of Typographic Styles_ are of that kind) -- so I am unsure, honestly, that extending frame syntax at its lowest level would be a good idea or not. 

I sincerely don't know if it's a "smarter" way or not -- page layout design is never simple (and Bringhurst remains quite general here[^1]) --, but my own approach to the topic in my resilient module was to propose a few (more-or-less standard) "page layouts," so that (end-)users wouldn't always have to dabble into the details of constructing their own low-level framesets. Ideally, these layouts can be extended (and I actually reluctantly added a "geometry" layout at one point for users who think they know what their margins should be). Notwithstanding implementation details, it does eventually looks like your function above (i.e. it ends up returning framesets).

[^1]: Or quite vague: "It is usual to set folios in the text size and to position them near the textblock" -- How to define "near"? ;) Well, let's be fair with Bringhurst, Tschichold is not very useful either to define how near it should be (beyond stating that it should obey harmony and proportions).

**Reply by jodros** [reply]
I know its unusual, but I'm sure it's a good idea!

> my resilient module

BTW I've read the parts of the documentation and code of your project, and I find interesting how some ideas are similar to mine, one in particular: modify the styling through a `toml` or `yaml` file. I think the SILE vanilla should have something like that, I'm gonna create another discussion to talk more about...

**Reply by Omikhleia** [reply]
>  I'm gonna create another discussion to talk more about...

Refers to #1946 

On the more specific question here regarding frame syntax, it's fairly clear the latter is already cluttered, and there are ways to address the concern on top of the existing frame definitions (as stated, with a layout layer wrapper) = Closing.
